### PR TITLE
Fixed issue with being unable to query with matcher using external label

### PR DESCRIPTION
**Pull request by Bplotka**
Signed-off-by: Bartek Plotka <bwplotka@gmail.com>

### CODE REVIEW

File: pkg/store/matchers.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+		default:
+			return "", errors.New("unknown matcher type")
+		}
+
+		b.WriteString(strconv.Quote(m.Value))
+	}
+
+	b.WriteByte('}')
+	return b.String(), nil
+}
```
Comment (fabxc) [inline pkg/store/matchers.go]: Missing newline, shouldn't gofmt fix this automatically?

File: pkg/store/promproxy.go [hunk 2]
```diff
@@ -156,6 +165,36 @@ func (p *PrometheusProxy) Series(ctx context.Context, r *storepb.SeriesRequest)
 	return res, nil
 }
 
+func extLabelsMatches(extLabels labels.Labels, ms []storepb.LabelMatcher) (bool, []storepb.LabelMatcher, error) {
+	elmap := map[string]string{}
+	for _, l := range extLabels {
+		elmap[l.Name] = l.Value
+	}
```
Comment (fabxc) [inline pkg/store/promproxy.go]: Why do we build this map? We have `labels.Labels` so `extLabels.Get(name) == ""` should be a sufficient check for non-existance.

In general for Prometheus an empty label is defined as equivalent to the label not existing. There's one exception that applies during federation – but that's more of a hack we can ignore in general.

File: pkg/store/promproxy_test.go [hunk 3]
```diff
@@ -112,3 +112,62 @@ func TestPrometheusProxy_LabelValues(t *testing.T) {
 
 	testutil.Equals(t, []string{"a", "b", "c"}, resp.Values)
 }
+
+func TestPrometheusProxy_Series_MatchExternalLabel(t *testing.T) {
+	p, err := testutil.NewPrometheus(":12347")
```
Comment (fabxc) [inline pkg/store/promproxy_test.go]: We should probably switch this to grab a random free port automatically and add a getter like `p.Addr()`. Bumping the port number by 1 for every test is probably not sustainable :D

**Comment by fabxc** [comment]
👍 in general

File: pkg/store/matchers.go (hunk 1 above)
Comment (Bplotka) [reply pkg/store/matchers.go]: Nope we need to add this to pre commit hook or something

Comment (Bplotka) [reply pkg/store/matchers.go]: I mean, yes - it fixes but I forgot to run it on this file ):

File: pkg/store/promproxy.go (hunk 2 above)
Comment (Bplotka) [reply pkg/store/promproxy.go]: sorry, forgot about `Get` method.

File: pkg/store/matchers.go (hunk 1 above)
Comment (fabxc) [reply pkg/store/matchers.go]: It should ideally be a save hook in your editor. `go fmt` is explicitly not intended as a pre-commit hook as I learned recently: https://tip.golang.org/doc/go1.10#gofmt

```
Note that these kinds of minor updates to gofmt are expected from time to time. In general, we recommend against building systems that check that source code matches the output of a specific version of gofmt. For example, a continuous integration test that fails if any code already checked into a repository is not “properly formatted” is inherently fragile and not recommended.
```

At least I interpret that if checking for formatting is discouraged, applying it automatically is so as well.

Comment (Bplotka) [reply pkg/store/matchers.go]: Good to know (: I have that under the special key, but not with the file save

File: pkg/testutil/prometheus.go [hunk 4]
```diff
@@ -59,11 +62,25 @@ func NewPrometheus(address string) (*Prometheus, error) {
...
+	if err != nil {
+		return 0, err
+	}
+
+	l, err := net.ListenTCP("tcp", addr)
+	if err != nil {
+		return 0, err
+	}
+	defer l.Close()
+	return l.Addr().(*net.TCPAddr).Port, nil
```
Comment (fabxc) [inline pkg/testutil/prometheus.go]: Mh, we still have a chance to lose the port again in between. But there's no real alternative I suppose. Should work well enough anyway.

We could pass Prometheus `:0` but I believe it doesn't expose the port it ended up with anywhere.

Comment (Bplotka) [reply pkg/testutil/prometheus.go]: Exactly that is the problem..

**Pull request merged**